axlearn/common/trainer.py

```diff
@@ -800,13 +805,15 @@ def restore_checkpoint(self, restore_step: Optional[int] = None) -> Optional[int
                         step,
                         restore_input_iter,
                     )
+                self._maybe_record_event(measurement.Event.END_DATA_LOADING)
```

Comment:
If this line got skipped due to self.checkpointer.restore() failure, the data loading end time will be missing for this event, is it expected?

Reply:
In that case, the badput would go into the "Unknown" bucket and not the "Data Loading" bucket. This is the expected behavior today, as there is currently no way to distinguish lost productive due to data loading if checkpointer.restore() failed since we don't really know if data loading even happened or not. We might however be able to put this into a "program runtime badput" if we can determine a better signal for data loading to isolate from infrastructure or hardware issues that could lead to "unknown" source of Badput as it stands today.